### DISCUSSION TITLE
Configurar build para Flutter usando Flavor

### DISCUSSION THREAD

**Discussion by jhonathanqz** [codemagic-ci-cd/codemagic-docs#2351] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

-

### Steps to reproduce
```
ios-workflow:
    name: iOS Workflow
    max_build_duration: 120
    integrations:
      app_store_connect: Accesys
    environment:
      groups:
        - api_credentials
        - google_credentials
      vars:
        STATIC_TOKEN: $CM_STATIC_TOKEN
        TEST: $CM_TEST
        STAGING: $CM_STAGING
        PRODUCTION: $CM_PRODUCTION
        URL: $CM_URL
      flutter: stable
    scripts:
      - name: Set up environment variables
        script: |
          echo "Valores antes do bloco de código:"
          echo "STAGING=$CM_STAGING" > .env
          echo "STATIC_TOKEN=$CM_STATIC_TOKEN" >> .env
          echo "PRODUCTION=$CM_PRODUCTION" >> .env
          echo "TEST=$CM_TEST" >> .env
          echo "URL=$CM_URL" >> .env
          echo "DELIMITER" >> .env
          echo "Valores depois do bloco de código:"
          cat .env
      - name: Set up code signing settings on Xcode project
        script: |
          xcode-project use-profiles
      - name: Get Flutter packages
        script: |
          flutter clean
          flutter pub get
      - name: Install pods
        script: |
          gem install cocoapods

          find . -name "Podfile" -execdir pod deintegrate \;
          find . -name "Podfile" -execdir pod install \;
          find . -name "Podfile" -execdir pod install --repo-update \;
      - name: Flutter build ipa
        script: |
          flutter clean
          flutter pub get

          FLAVORS_FILE="flavors_ios.txt"
          FLAVORS=$(cat $FLAVORS_FILE)

          IFS=',' read -ra FLAVORS_ARRAY <<< "$FLAVORS"
          echo "Found flavors: ${FLAVORS_ARRAY[@]}"

          for FLAVOR_WITH_ID in "${FLAVORS_ARRAY[@]}"; do
            IFS=' ' read -ra FLAVOR_DATA <<< "$FLAVOR_WITH_ID"
            FLAVOR=${FLAVOR_DATA[0]}
            APP_ID=${FLAVOR_DATA[1]}
            BUNDLE_IDENTIFIER=${FLAVOR_DATA[2]}

            echo "**Running iOS Workflow for flavor: $FLAVOR with appId: $APP_ID and bundle_identifier: $BUNDLE_IDENTIFIER"

            LATEST_BUILD_NUMBER=$(app-store-connect get-latest-app-store-build-number "$APP_ID")
            echo "**LATEST_BUILD_NUMBER: $LATEST_BUILD_NUMBER"

            NEXT_BUILD_NUMBER=$((LATEST_BUILD_NUMBER + 1))
            echo "**NEXT_BUILD_NUMBER: $NEXT_BUILD_NUMBER"

            flutter build ipa --flavor $FLAVOR -t lib/main-$FLAVOR.dart --release \
              --build-name=1.0.0 \
              --build-number=$NEXT_BUILD_NUMBER \
              --export-options-plist=/Users/builder/export_options.plist \
              --dart-define=FLUTTER_APP_BUNDLE_IDENTIFIER=$BUNDLE_IDENTIFIER
              
          done
    artifacts:
      - build/ios/ipa/*.ipa
      - /tmp/xcodebuild_logs/*.log
      - flutter_drive.log
    publishing:
      email:
        recipients:
          - jhonathanqz011@gmail.com
        notify:
          success: true
          failure: true
      app_store_connect:
        auth: integration

        submit_to_testflight: true
        submit_to_app_store: true

```

### Expected results

Preciso que ele gere um .ipa para cada Flavor mencionado no arquivos flavors_ios.txt e logo em seguida publique todos na apple_store.

A mesma ideia funcionou para o Android, porém para iOS estou recebendo um erro e não estou conseguindo resolver.

### Actual results

O codemagic está me retornando o seguinte erro:

**Running iOS Workflow for flavor: facility with appId: <app_id>
Found build number 2 from App Store version v1.1.0(2)
**LATEST_BUILD_NUMBER: 2
**NEXT_BUILD_NUMBER: 3
Could not find an option with short name "-a".

Run 'flutter -h' (or 'flutter <command> -h') for available flutter commands and options.

Build failed :|
Step 7 script `Flutter build ipa` exited with status code 64

### Build id (optional)

65bd213c105211fce0fefa5e

**Comment by icarusdust** [comment]
Hi @jhonathanqz, the problem is that you are adding `allowProvisioningUpdates` to `flutter build ipa`, and the command does not recognise it. All you need is to remove the flag from the build command.

**Reply by jhonathanqz** [reply]
Mesmo assim estoura outro erro @icarusdust .
Segue mais informações do erro:

Obs: Todos os profile eu já importei nas configurações do time no codemagic.
```
Xcode archive done.                                          5.9s
Failed to build iOS app
Error (Xcode): No profiles for 'br.com.accesys.lojasExpress.facilityHomeMarket' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'br.com.accesys.lojasExpress.facilityHomeMarket'. Automatic signing is disabled and unable to generate a profile. To enable automatic signing, pass -allowProvisioningUpdates to xcodebuild.
/Users/builder/clone/ios/Runner.xcodeproj


It appears that there was a problem signing your application prior to installation on the device.

Verify that the Bundle Identifier in your project is your signing id in Xcode
  open ios/Runner.xcworkspace

Also try selecting 'Product > Build' to fix the problem.
Encountered error while archiving for device.


Build failed :|
Step 7 script `Flutter build ipa` exited with status code 1
```

ID deste último build: 65c660307ae3e61db812e6db

Meu codemagic está atualizado no tópico dessa discussão.

**Reply by icarusdust** [reply]
So, removing the flag helped you resolve the initial error. 

The current error is very much expected because you are not exporting the provisioning profile from your team settings to `codemagic.yaml`. Therefore, we can safely say that your code signing settings are not in place, hence why the error.
 
What you need to do is to add the following lines in YAML:

```
    environment:
      ios_signing:
        distribution_type: app_store # or: ad_hoc | development | enterprise
        bundle_identifier: br.com.accesys.lojasExpress.facilityHomeMarket
```

**Reply by jhonathanqz** [reply]
@icarusdust entao. Mas esse workflow está fazendo o build de vários Flavors do meu projeto Flutter. Tem como eu colocar mais de um bundle_identifier no workflow? 

Exemplo: ele vai gerar o build de 10 aplicativos, quero que ele já publique os 10. Cada um com seu bundle_identifier

**Reply by jhonathanqz** [reply]
@icarusdust Alguma sugestão?

**Reply by icarusdust** [reply]
@jhonathanqz you should build and publish the builds separately with different flavors that hold different bundle identifiers along with each's separate provisioning profiles. My question would be how do you achieve it when building the project on your local machine ?

**Reply by jhonathanqz** [reply]
@icarusdust  em minha máquina eu publico um de cada vez, por isso queria automatizar essa parte da publicação, pois meu projeto é um aplicativo white label, com 30 Flavor. Então queria automatizar para ele já fazer o processo de build dos 30 e já publicar automaticamente. Fazendo essa mesma logica com o android deu certo, porém com a apple estou tendo problemas. Você fala então, que se eu tenho 30 flavors, eu vou precisar de 30 workflow do tipo iOS para cada compilar os flavors então?

**Reply by icarusdust** [reply]
First off, worth pointing out that building flavors and white-labeling apps should be regarded as separate topics.

If you white-label your apps, then you can turn to REST Api which will be used to trigger builds for multiple clients simultaneously by looping through client IDs considering their bundle identifiers, provisioning profiles, assets etc. In essence, one workflow will be used by modifying the appropriate steps of it based on your clients' info provided.

To begin with, I suggest you check [this documentation](https://docs.codemagic.io/yaml-quick-start/white-label-getting-started/) where you can find out more about how to trigger builds, how to set up a proper white-labeling structure etc.

**Reply by jhonathanqz** [reply]
@icarusdust Muito obrigado!!! Vou verificar e tentar aplicar.
